File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.cc
#include<iostream>
#include<unordered_map>
using namespace std;

int main() {
    
    // 입력
    int T;
    cin >> T;

    while (T-- > 0) {

        string str;
        int K;
        cin >> str >> K;

        // 변수 초기화
        unordered_map<char, int> count;
        for (int i = 0; i < str.size(); i++) {
            count[str[i]]++;  // char의 아스키코드 이용 (a:97, b:98 ...)
        }

        
        // K개의 정수가 포함된 길이 구하기
        int short_length = 10001, long_length = -1;
        for (int i = 0; i < str.size(); i++) {
            
            // K개보다 적은 경우
            if (count[str[i]] < K) {
                continue;
            }

            int curCount = 0;
            for (int j = i; j < str.size(); j++) {

                if (str[i] == str[j]) {
                    curCount++;
                }

                // 정확히 K개일때
                if (curCount == K) {
                    short_length = min(short_length, j - i + 1);
                    long_length = max(long_length, j - i + 1);
                    break;
                }
            }

        }


        // 출력
        if (short_length == 10001) {
            cout << "-1\n";
        }
        else {
            cout << short_length << " " << long_length << "\n";
        }
    }
}

File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.java
import java.io.*;
import java.util.*;

// 10:50
class Main {
	
	static int num;
	static String str;
	static int min, max;
	
	public static void main(String[] args) throws Exception {
		/* 입력 */
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int tc = Integer.parseInt(br.readLine());
		
		while(tc-- > 0) {
			str = br.readLine();
			num = Integer.parseInt(br.readLine());
			
			getStringLength();			
			
			/* 출력 */
			if (min == Integer.MAX_VALUE) {
				System.out.println(-1);
			}
			else {
				System.out.println(min + " " + max);
			}
		}
	}
	
	/* 가장 짧은/긴 문자열 길이 구하기 */
	private static void getStringLength() {
		int[] count = getAlphabetCount();
		min = Integer.MAX_VALUE;
		max = Integer.MIN_VALUE;
		for (int i = 0; i < str.length(); i++) {

			// 현재 문자의 개수가 num보다 클 때
			if (count[str.charAt(i) - 'a'] >= num) {
				int curCount = 0; // 현재 문자의 개수
				for (int j = i; j < str.length(); j++) {
					
					// 첫번째 문자와 마지막 문자가 같을때
					if (str.charAt(i) == str.charAt(j)) {
						curCount++;
						
						if (curCount == num) {
							min = Math.min(min,  j - i + 1);
							max = Math.max(max,  j - i + 1);
							break;
						}
						
					}
				}
			}			
		}
		
	}

	private static int[] getAlphabetCount() {
		int[] count = new int[26];
		
		for (char s : str.toCharArray()) {
			count[s - 'a']++;
		}
		return count;
	}
}